Update foreach binary ops with a single scalar and list of tensors

izdeby opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52450 wip
* #49039 [wip] Replace optimizers in torch.optim with the ones from torch.optim._multi_tensor
* #52449 Add type promotion to a fast path for foreach APIs
* #52448 Support broadcasting via slow path
* #48224 [wip] Pushed unsupported scenarios to slow path
* #51170 Updated foreach pointwise ops to support complex and bloat16 on CUDA
* #49714 Update foreach max/min
* #51060 Refactor foreach pointwise ops tests to use OpInfo
* #51059 Refactor foreach binary ops tests with tensor lists to use OpInfo
* #51058 Refactor foreach binary ops tests with scalars to use OpInfo
* #49712 Refactor foreach unary ops tests to use OpInfo
* **#49250 Update foreach binary ops with a single scalar and list of tensors**
* #49249 Update foreach binary ops with a scalar list
* #51131 Refactor ForeachUtils.h

Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

---- 
Update logic for 2 sets of the foreach APIs:
1) Binary ops with a scalar
2) Binary ops with a list of tensors
- Update foreach_sub error checking
- Update division logic to check for type promotion
- Updated tests

Commit:
Add division logic to a slow/fast path

[ghstack-poisoned]

Commit:
Update on "Add division logic to a slow/fast path"


Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

[ghstack-poisoned]

Commit:
Update on "Add division logic to a slow/fast path"


Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

[ghstack-poisoned]

Commit:
Update on "[wip] Add division logic to a slow/fast path"


Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

[ghstack-poisoned]

Commit:
Update on "[wip] Add division logic to a slow/fast path"


Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

[ghstack-poisoned]

Commit:
Update on "Update ForeachBinaryOpScalar.cu"


Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

[ghstack-poisoned]

Commit:
Update on "Update ForeachBinaryOpScalar.cu"


Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

[ghstack-poisoned]

Commit:
Update on "Update ForeachBinaryOpScalar.cu"


Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

---- 
- Update ForeachBinaryOpScalar.cu to throw better errors in case of sub 
- Update division logic to check for type promotion

[ghstack-poisoned]

Commit:
Update on "Update ForeachBinaryOpScalar.cu"


Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

---- 
- Update ForeachBinaryOpScalar.cu to throw better errors in case of sub 
- Update division logic to check for type promotion

[ghstack-poisoned]

Commit:
Update on "Update ForeachBinaryOpScalar.cu"


Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

---- 
- Update ForeachBinaryOpScalar.cu to throw better errors in case of sub 
- Update division logic to check for type promotion

[ghstack-poisoned]

Commit:
Update on "Update ForeachBinaryOpScalar.cu"


Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

---- 
- Update ForeachBinaryOpScalar.cu to throw better errors in case of sub 
- Update division logic to check for type promotion

[ghstack-poisoned]

Commit:
Update on "Update ForeachBinaryOpScalar.cu"


Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

---- 
- Update ForeachBinaryOpScalar.cu to throw better errors in case of sub 
- Update division logic to check for type promotion

[ghstack-poisoned]

Commit:
Update on "Update ForeachBinaryOpScalar.cu"


Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

---- 
- Update ForeachBinaryOpScalar.cu to throw better errors in case of sub 
- Update division logic to check for type promotion

[ghstack-poisoned]

Commit:
Update on "Update ForeachBinaryOpScalar.cu and ForeachBinaryOpList.cu"


Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

---- 
- Update ForeachBinaryOpScalar.cu to throw better errors in case of sub 
- Update division logic to check for type promotion

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/native/cuda/ForeachBinaryOpList.cu:
```diff
@@ -89,7 +89,7 @@ std::vector<Tensor> foreach_tensor_##NAME##_list_kernel_cuda(TensorList tensors1
 FOREACH_BINARY_OP_LIST_ALPHA(sub, std::minus);
-FOREACH_BINARY_OP_LIST(mul, std::multiplies);
-FOREACH_BINARY_OP_LIST(div, std::divides);
+FOREACH_BINARY_OP_LIST(mul, std::multiplies, false);
+FOREACH_BINARY_OP_LIST(div, std::divides, true);
```
nit: adding a comment in front of these make them more readable: `/*division_op*/false`

There will probably be more binary operations that do int->float promotion in the future, but I suppose we can refactor when they come along

zou3519 commented on aten/src/ATen/native/cuda/ForeachBinaryOpScalar.cu:
```diff
@@ -58,16 +58,53 @@ void foreach_tensor_##NAME##_scalar_kernel_cuda_(TensorList tensors, Scalar scal
+void foreach_tensor_sub_scalar_kernel_cuda_(TensorList tensors, Scalar scalar) {
+    check_foreach_api_restrictions(tensors);
+
+    TORCH_CHECK(tensors[0].scalar_type() != kBool || !scalar.isBoolean(),
+              "Subtraction, the `-` operator, with two bool tensors is not supported."
+              "Use the `^` or `logical_xor()` operator instead.")
+    TORCH_CHECK(tensors[0].scalar_type() != kBool && !scalar.isBoolean(),
+              "Subtraction, the `-` operator, with a bool tensor is not supported. "
+              "If you are trying to invert a mask, use the `~` or `logical_not()` operator instead.");
```
As I mentioned in the last PR in the stack, we should really refactor these messages out so that they're not being repeated so much

zou3519 commented on test/test_foreach.py:
```diff
@@ -341,34 +341,65 @@ def test_int_scalar(self, device, dtype):
-                expected = [torch_bin_op(t, scalar) for t in tensors]
 
-                res = foreach_bin_op(tensors, scalar)
+                if dtype == torch.bool and foreach_bin_op == foreach_bin_op == torch._foreach_sub:
```
`foreach_bin_op == foreach_bin_op == torch._foreach_sub`? Occurs on the next line too

zou3519 commented on test/test_foreach.py:
```diff
@@ -341,34 +341,65 @@ def test_int_scalar(self, device, dtype):
                         foreach_bin_op_(tensors, scalar)
-                    return
 
+                    with self.assertRaisesRegex(RuntimeError, "can't be cast to the desired output type"):
+                        [t.div_(scalar) for t in tensors]
+
+                    with self.assertRaisesRegex(RuntimeError, "can't be cast to the desired output type"):
+                        [t.mul_(scalar) for t in tensors]
+
+                    with self.assertRaisesRegex(RuntimeError, "can't be cast to the desired output type"):
+                        [t.add_(scalar) for t in tensors]
 
+                    with self.assertRaisesRegex(RuntimeError, "Subtraction, the `-` operator, with a bool"):
+                        [t.sub_(scalar) for t in tensors]
```
I'm confused -- what is being tested here?

zou3519 commented on test/test_foreach.py:
```diff
@@ -341,34 +341,65 @@ def test_int_scalar(self, device, dtype):
-                        foreach_bin_op_(tensors, scalar)
-                    return
+                    with self.assertRaisesRegex(RuntimeError, "can't be cast to the desired output type"):
+                        [t.div_(scalar) for t in tensors]
 
-                # TODO[type promotion]: Fix once type promotion is enabled.
-                if dtype in torch.testing.integral_types() and self.device_type == 'cuda':
-                    self.assertEqual(res, [e.to(dtype) for e in expected])
+                    with self.assertRaisesRegex(RuntimeError, "can't be cast to the desired output type"):
+                        torch._foreach_div_(tensors, scalar)
+                    continue
```
Why are we checking for `self.device_type == "cpu"` ? This behavior should happen for both CUDA and CPU

zou3519 commented on test/test_foreach.py:
```diff
@@ -341,34 +341,65 @@ def test_int_scalar(self, device, dtype):
-                    self.assertEqual(res, expected)
+                self.assertEqual(res, expected)
+
+                # In case of In-place op, we can't change the dtype
+                if (expected[0].dtype == dtype):
                     foreach_bin_op_(tensors, scalar)
                     self.assertEqual(tensors, expected)
+                else:
+                    with self.assertRaisesRegex(RuntimeError, "can't be cast to the desired output type"):
+                        foreach_bin_op_(tensors, scalar)
```
Is this check only needed for the foreach inplace division op?

zou3519 commented on test/test_foreach.py:
```diff
@@ -583,46 +614,27 @@ def test_bool_scalar(self, device, dtype):
-                            foreach_bin_op_(tensors, scalar)
-                            self.assertEqual(tensors, res)
-                    else:
+                if dtype in torch.testing.integral_types_and(torch.bool) and foreach_bin_op == foreach_bin_op == torch._foreach_div:
```
"foreach_bin_op == foreach_bin_op" lol

izdeby committed:
Update on "Update ForeachBinaryOpScalar.cu and ForeachBinaryOpList.cu"


Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

---- 
- Update ForeachBinaryOpScalar.cu to throw better errors in case of sub 
- Update division logic to check for type promotion

[ghstack-poisoned]

Commit:
Update on "Update ForeachBinaryOpScalar.cu and ForeachBinaryOpList.cu"


Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

---- 
- Update ForeachBinaryOpScalar.cu to throw better errors in case of sub 
- Update division logic to check for type promotion

[ghstack-poisoned]

Commit:
Update on "Update ForeachBinaryOpScalar.cu and ForeachBinaryOpList.cu"


Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

---- 
- Update ForeachBinaryOpScalar.cu to throw better errors in case of sub 
- Update division logic to check for type promotion

[ghstack-poisoned]

Commit:
Update on "Update ForeachBinaryOpScalar.cu and ForeachBinaryOpList.cu"


Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

---- 
- Update ForeachBinaryOpScalar.cu to throw better errors in case of sub 
- Update division logic to check for type promotion

[ghstack-poisoned]

Commit:
Update on "Update ForeachBinaryOpScalar.cu and ForeachBinaryOpList.cu"


Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

---- 
- Update ForeachBinaryOpScalar.cu to throw better errors in case of sub 
- Update division logic to check for type promotion

[ghstack-poisoned]

Commit:
Update on "Update foreach_div and foreach_sub logic"


Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

---- 
- Update foreach_sub error checkingz
- Update division logic to check for type promotion
- Updated tests

[ghstack-poisoned]

Commit:
Update on "Update foreach_div and foreach_sub logic"


Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

---- 
- Update foreach_sub error checking
- Update division logic to check for type promotion
- Updated tests

[ghstack-poisoned]

Commit:
Update on "Update foreach_div and foreach_sub logic"


Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

---- 
- Update foreach_sub error checking
- Update division logic to check for type promotion
- Updated tests

[ghstack-poisoned]

Commit:
Update on "Update foreach_div and foreach_sub logic"


Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

---- 
- Update foreach_sub error checking
- Update division logic to check for type promotion
- Updated tests

[ghstack-poisoned]

Commit:
Update on "Update foreach_div and foreach_sub logic"


Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

---- 
- Update foreach_sub error checking
- Update division logic to check for type promotion
- Updated tests

[ghstack-poisoned]

Commit:
Update on "Update foreach_div and foreach_sub logic"


Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

---- 
- Update foreach_sub error checking
- Update division logic to check for type promotion
- Updated tests

[ghstack-poisoned]

Commit:
Update on "Update foreach_div and foreach_sub logic"


Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

---- 
- Update foreach_sub error checking
- Update division logic to check for type promotion
- Updated tests

[ghstack-poisoned]

izdeby committed:
Update on "Update foreach_div and foreach_sub logic"


Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

---- 
- Update foreach_sub error checking
- Update division logic to check for type promotion
- Updated tests

[ghstack-poisoned]

Commit:
Update on "Update foreach_div and foreach_sub logic"


Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

---- 
- Update foreach_sub error checking
- Update division logic to check for type promotion
- Updated tests

[ghstack-poisoned]

Commit:
Update on "Update foreach binary ops with a single scalar and list of tensors"


Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

---- 
Update logic for 2 set of foreach APIs:
1) Binary ops with a scalar
2) Binary ops with a list of tensors
- Update foreach_sub error checking
- Update division logic to check for type promotion
- Updated tests

[ghstack-poisoned]

Commit:
Update on "Update foreach binary ops with a single scalar and list of tensors"


Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

---- 
Update logic for 2 set of foreach APIs:
1) Binary ops with a scalar
2) Binary ops with a list of tensors
- Update foreach_sub error checking
- Update division logic to check for type promotion
- Updated tests

[ghstack-poisoned]

Commit:
Update on "Update foreach binary ops with a single scalar and list of tensors"


Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

---- 
Update logic for 2 sets of the foreach APIs:
1) Binary ops with a scalar
2) Binary ops with a list of tensors
- Update foreach_sub error checking
- Update division logic to check for type promotion
- Updated tests

[ghstack-poisoned]

Commit:
Update on "Update foreach binary ops with a single scalar and list of tensors"


Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

---- 
Update logic for 2 sets of the foreach APIs:
1) Binary ops with a scalar
2) Binary ops with a list of tensors
- Update foreach_sub error checking
- Update division logic to check for type promotion
- Updated tests

[ghstack-poisoned]

Commit:
Update on "Update foreach binary ops with a single scalar and list of tensors"


Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

---- 
Update logic for 2 sets of the foreach APIs:
1) Binary ops with a scalar
2) Binary ops with a list of tensors
- Update foreach_sub error checking
- Update division logic to check for type promotion
- Updated tests

[ghstack-poisoned]

Commit:
Update on "Update foreach binary ops with a single scalar and list of tensors"


Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

---- 
Update logic for 2 sets of the foreach APIs:
1) Binary ops with a scalar
2) Binary ops with a list of tensors
- Update foreach_sub error checking
- Update division logic to check for type promotion
- Updated tests

[ghstack-poisoned]

Commit:
Update on "Update foreach binary ops with a single scalar and list of tensors"


Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

---- 
Update logic for 2 sets of the foreach APIs:
1) Binary ops with a scalar
2) Binary ops with a list of tensors
- Update foreach_sub error checking
- Update division logic to check for type promotion
- Updated tests

[ghstack-poisoned]

Commit:
Update on "Update foreach binary ops with a single scalar and list of tensors"


Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

---- 
Update logic for 2 sets of the foreach APIs:
1) Binary ops with a scalar
2) Binary ops with a list of tensors
- Update foreach_sub error checking
- Update division logic to check for type promotion
- Updated tests

[ghstack-poisoned]

izdeby committed:
Update on "Update foreach binary ops with a single scalar and list of tensors"


Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

---- 
Update logic for 2 sets of the foreach APIs:
1) Binary ops with a scalar
2) Binary ops with a list of tensors
- Update foreach_sub error checking
- Update division logic to check for type promotion
- Updated tests

[ghstack-poisoned]

zou3519 commented on test/test_foreach.py:
```diff
@@ -353,34 +353,60 @@ def test_int_scalar(self, device, dtype):
-                res = foreach_bin_op(tensors, scalar)
 
                 if dtype == torch.bool:
+                    if foreach_bin_op == torch._foreach_sub:
```
nit: this test is called `test_int_scalar`. That's not very descriptive, it looks like what is being tested here is binary ops where the second argument is an int. Maybe something like `test_tensorlist_scalar_op`?

zou3519 commented on test/test_foreach.py:
```diff
@@ -353,34 +353,60 @@ def test_int_scalar(self, device, dtype):
+                self.assertEqual(res, expected)
+
+                # In case of In-place op, we can't change the dtype
+                foreach_bin_op_(tensors, scalar)
```
NB: this modifies the input "tensors" in-place, which means that the values are different the next iteration of the outer for-loop. But that seems to be okay. Usually to avoid mutating we would clone the input before performing the in-place operation.

No action item required here. It may possibly be nicer if we cloned `tensors` before calling the in-place op but that's subjective. I also see this pattern in the other test that is being modified in this PR.

zou3519 reviewed (approved):
Had some minor comments. Splitting up the PRs really helped, this was easy to review

izdeby committed:
Update on "Update foreach binary ops with a single scalar and list of tensors"


Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

---- 
Update logic for 2 sets of the foreach APIs:
1) Binary ops with a scalar
2) Binary ops with a list of tensors
- Update foreach_sub error checking
- Update division logic to check for type promotion
- Updated tests

[ghstack-poisoned]

izdeby committed:
Update on "Update foreach binary ops with a single scalar and list of tensors"


Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

---- 
Update logic for 2 sets of the foreach APIs:
1) Binary ops with a scalar
2) Binary ops with a list of tensors
- Update foreach_sub error checking
- Update division logic to check for type promotion
- Updated tests

[ghstack-poisoned]

izdeby committed:
Update on "Update foreach binary ops with a single scalar and list of tensors"


Differential Revision: [D25502938](https://our.internmc.facebook.com/intern/diff/D25502938)

---- 
Update logic for 2 sets of the foreach APIs:
1) Binary ops with a scalar
2) Binary ops with a list of tensors
- Update foreach_sub error checking
- Update division logic to check for type promotion
- Updated tests

[ghstack-poisoned]
